File: tasks/python/task-03-ticket-search-empty/solution/tickets.py
"""Look up ticket ids in a sorted list.

The list holds integer ticket ids in ascending order.
"""


def find_ticket(ticket_ids, ticket_id):
    """Return the index of `ticket_id` in the sorted list, or -1."""
    if not ticket_ids:
        return -1
    if ticket_id < ticket_ids[0] or ticket_id > ticket_ids[-1]:
        return -1
    low = 0
    high = len(ticket_ids) - 1
    while low <= high:
        middle = (low + high) // 2
        if ticket_ids[middle] == ticket_id:
            return middle
        if ticket_ids[middle] < ticket_id:
            low = middle + 1
        else:
            high = middle - 1
    return -1


def has_ticket(ticket_ids, ticket_id):
    """Return True if the id is in the list."""
    return find_ticket(ticket_ids, ticket_id) != -1


def missing_tickets(ticket_ids, wanted):
    """Return the wanted ids that are not in the sorted list."""
    return [ticket_id for ticket_id in wanted if not has_ticket(ticket_ids, ticket_id)]


def add_ticket(ticket_ids, ticket_id):
    """Insert the id and keep the list sorted. Ignore duplicates."""
    if has_ticket(ticket_ids, ticket_id):
        return
    ticket_ids.append(ticket_id)
    ticket_ids.sort()

File: tasks/typescript/task-03-ticket-search-empty/solution/tickets.ts
/**
 * Look up tickets in a list that is sorted by id.
 *
 * Each ticket has an integer id and a title. The ids are in ascending order.
 */

export type Ticket = {
  id: number;
  title: string;
};

/** Return the index of the ticket with this id in the sorted list, or -1. */
export function findTicket(tickets: Ticket[], id: number): number {
  if (tickets.length === 0) {
    return -1;
  }
  if (id < tickets[0].id || id > tickets[tickets.length - 1].id) {
    return -1;
  }
  let low = 0;
  let high = tickets.length - 1;
  while (low <= high) {
    const middle = Math.floor((low + high) / 2);
    if (tickets[middle].id === id) {
      return middle;
    }
    if (tickets[middle].id < id) {
      low = middle + 1;
    } else {
      high = middle - 1;
    }
  }
  return -1;
}

/** Return true if a ticket with this id is in the list. */
export function hasTicket(tickets: Ticket[], id: number): boolean {
  return findTicket(tickets, id) !== -1;
}

/** Return the wanted ids that have no ticket in the sorted list. */
export function missingTickets(tickets: Ticket[], wanted: number[]): number[] {
  return wanted.filter((id) => !hasTicket(tickets, id));
}

/** Insert the ticket and keep the list sorted by id. Ignore duplicate ids. */
export function addTicket(tickets: Ticket[], ticket: Ticket): void {
  if (hasTicket(tickets, ticket.id)) {
    return;
  }
  tickets.push(ticket);
  tickets.sort((a, b) => a.id - b.id);
}
